### ISSUE TITLE
Lowering operation for reconstruction

### ISSUE THREAD

**Issue by modichirag** [DifferentiableUniverseInitiative/flowpm#35]
The time for lowering operation for reconstruction graph with N-Body/PM forward model is too high. 
It is also high for LPT forward model, but not unreasonably so. 

The reconstruction graph is same as the forward model graph with 2 extra computes - gradient with respect to the linear field and update operation for the linear field.
In the commit - https://github.com/modichirag/flowpm/commit/5c3fe394be915a3e462fab41d221ea3eeee912df
it can be found in the file- https://github.com/modichirag/flowpm/blob/recon/examples/pyramid_recon.py
(in case future commits change the graph)

For timing comparison - 
Config:  1 node, 8 GPUs on cori with nx*ny=4x2 config of mesh, 64^3 PM grid, 5 step PM model and pyramid scheme
- time to lowering a graph of forward nbody model with the same number of steps is ~60 second
- time to lowering a graph of recon i.e. forward nbody model + gradients wrt linear field is ~2700 second
- In case of single-mesh scheme, time to lowering a graph of recon i.e. forward nbody model + gradients wrt linear field is ~1100 second

This issue is important since it will make it hard to experiment with N-body models for reconstruction.

**Comment by EiffL** [comment]
So, I agree it's an inconvenience, but two comments:
  - This wouldn't happen if we had an SIMD implementation for GPUs, or if we run on TPUs.
  - For some applications, this graph setup steps will be smaller than the actual training time.

I'm not sure how to speed up the graph building. but one thing is certain, the more compute devices, the heavier this is going to become with the device placement mesh implementation

**Comment by EiffL** [comment]
almost there.  This should be solved when we switch to horovod communications

**Comment by EiffL** [comment]
This is  a non issue with the new backend for mesh tensorflow
